File: Runtime/CoreLib.Script/String.js
///////////////////////////////////////////////////////////////////////////////
// String Extensions

ss.splitWithCharsAndSplitOptions = function#? DEBUG ss$splitWithCharsAndSplitOptions##(s, chars, options) {
	var usedOption = options || 0;
	var result = s.split(new RegExp("[" + String.fromCharCode.apply(null, chars) + "]"));
	if(usedOption === 0)
		return result;

	var parts = [];
	for (var i in result)
		if(result[i].length > 0)
			parts.push(result[i]);

	return parts;
};

ss.compareStrings = function#? DEBUG ss$compareStrings##(s1, s2, ignoreCase) {
	if (ignoreCase) {
		if (s1) {
			s1 = s1.toUpperCase();
		}
		if (s2) {
			s2 = s2.toUpperCase();
		}
	}
	s1 = s1 || '';
	s2 = s2 || '';

	if (s1 == s2) {
		return 0;
	}
	if (s1 < s2) {
		return -1;
	}
	return 1;
};

ss.endsWithString = function#? DEBUG ss$endsWithString##(s, suffix) {
	if (!suffix.length) {
		return true;
	}
	if (suffix.length > s.length) {
		return false;
	}
	return (s.substr(s.length - suffix.length) == suffix);
};

ss._formatString = function#? DEBUG ss$_formatString##(format, values, useLocale) {
	if (!ss._formatRE) {
		ss._formatRE = /(\{[^\}^\{]+\})/g;
	}

	return format.replace(ss._formatRE,
						  function(str, m) {
							  var index = parseInt(m.substr(1));
							  var value = values[index + 1];
							  if (ss.isNullOrUndefined(value)) {
								  return '';
							  }
							  if (ss.isInstanceOfType(value, ss_IFormattable)) {
								  var formatSpec = null;
								  var formatIndex = m.indexOf(':');
								  if (formatIndex > 0) {
									  formatSpec = m.substring(formatIndex + 1, m.length - 1);
								  }
								  return ss.format(value, formatSpec);
							  }
							  else {
								  return useLocale ? value.toLocaleString() : value.toString();
							  }
						  });
};

ss.formatString = function#? DEBUG String$format##(format) {
	return ss._formatString(format, arguments, /* useLocale */ false);
};

ss.stringFromChar = function#? DEBUG ss$stringFromChar##(ch, count) {
	var s = ch;
	for (var i = 1; i < count; i++) {
		s += ch;
	}
	return s;
};

ss.htmlDecode = function#? DEBUG ss$htmlDecode##(s) {
	return s.replace(/&([^;]+);/g, function(_, e) {
		if (e[0] === '#')
			return String.fromCharCode(parseInt(e.substr(1)));
		switch (e) {
			case 'quot': return '"';
			case 'apos': return "'";
			case 'amp': return '&';
			case 'lt': return '<';
			case 'gt': return '>';
			default : return '&' + e + ';';
		}
	});
};

ss.htmlEncode = function#? DEBUG ss$htmlEncode##(s) {
	return s.replace(/&/g, '&amp;').replace(/"/g, '&quot;').replace(/'/g, '&#39;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
};

ss.jsEncode = function#? DEBUG ss$jsEncode##(s, q) {
	s = s.replace(/\\/g, '\\\\').replace(/'/g, "\\'").replace(/"/g, '\\"');
	return q ? '"' + s + '"' : s;
};

ss.indexOfAnyString = function#? DEBUG ss$indexOfAnyString##(s, chars, startIndex, count) {
	var length = s.length;
	if (!length) {
		return -1;
	}

	chars = String.fromCharCode.apply(null, chars);
	startIndex = startIndex || 0;
	count = count || length;

	var endIndex = startIndex + count - 1;
	if (endIndex >= length) {
		endIndex = length - 1;
	}

	for (var i = startIndex; i <= endIndex; i++) {
		if (chars.indexOf(s.charAt(i)) >= 0) {
			return i;
		}
	}
	return -1;
};

ss.insertString = function#? DEBUG ss$insertString##(s, index, value) {
	if (!value) {
		return s;
	}
	if (!index) {
		return value + s;
	}
	var s1 = s.substr(0, index);
	var s2 = s.substr(index);
	return s1 + value + s2;
};

ss.isNullOrEmptyString = function#? DEBUG ss$isNullOrEmptyString##(s) {
	return !s || !s.length;
};

ss.lastIndexOfAnyString = function#? DEBUG ss$lastIndexOfAnyString##(s, chars, startIndex, count) {
	var length = s.length;
	if (!length) {
		return -1;
	}

	chars = String.fromCharCode.apply(null, chars);
	startIndex = startIndex || length - 1;
	count = count || length;

	var endIndex = startIndex - count + 1;
	if (endIndex < 0) {
		endIndex = 0;
	}

	for (var i = startIndex; i >= endIndex; i--) {
		if (chars.indexOf(s.charAt(i)) >= 0) {
			return i;
		}
	}
	return -1;
};

ss.localeFormatString = function#? DEBUG ss$localeFormatString##(format) {
	return ss._formatString(format, arguments, /* useLocale */ true);
};

ss.padLeftString = function#? DEBUG ss$padLeftString##(s, totalWidth, ch) {
	if (s.length < totalWidth) {
		ch = String.fromCharCode(ch || 0x20);
		return ss.stringFromChar(ch, totalWidth - s.length) + s;
	}
	return s;
};

ss.padRightString = function#? DEBUG ss$padRightString##(s, totalWidth, ch) {
	if (s.length < totalWidth) {
		ch = String.fromCharCode(ch || 0x20);
		return s + ss.stringFromChar(ch, totalWidth - s.length);
	}
	return s;
};

ss.removeString = function#? DEBUG ss$removeString##(s, index, count) {
	if (!count || ((index + count) > this.length)) {
		return s.substr(0, index);
	}
	return s.substr(0, index) + s.substr(index + count);
};

ss.replaceAllString = function#? DEBUG ss$replaceAllString##(s, oldValue, newValue) {
	newValue = newValue || '';
	return s.split(oldValue).join(newValue);
};

ss.startsWithString = function#? DEBUG ss$startsWithString##(s, prefix) {
	if (!prefix.length) {
		return true;
	}
	if (prefix.length > s.length) {
		return false;
	}
	return (s.substr(0, prefix.length) == prefix);
};

if (!String.prototype.trim) {
	String.prototype.trim = function#? DEBUG String$trim##() {
		return ss.trimStartString(ss.trimEndString(this));
	};
}

ss.trimEndString = function#? DEBUG ss$trimEndString##(s, chars) {
	return s.replace(chars ? new RegExp('[' + String.fromCharCode.apply(null, chars) + ']+$') : /\s*$/, '');
};

ss.trimStartString = function#? DEBUG ss$trimStartString##(s, chars) {
	return s.replace(chars ? new RegExp('^[' + String.fromCharCode.apply(null, chars) + ']+') : /^\s*/, '');
};

ss.trimString = function#? DEBUG ss$trimString##(s, chars) {
	return ss.trimStartString(ss.trimEndString(s, chars), chars);
};

ss.lastIndexOfString = function#? DEBUG ss$lastIndexOfString##(s, search, startIndex, count) {
	var index = s.lastIndexOf(search, startIndex);
	return (index < (startIndex - count + 1)) ? -1 : index;
};

ss.indexOfString = function#? DEBUG ss$indexOfString##(s, search, startIndex, count) {
	var index = s.indexOf(search, startIndex);
	return ((index + search.length) <= (startIndex + count)) ? index : -1;
};

File: Runtime/CoreLib/String.cs
// String.cs
// Script#/Libraries/CoreLib
// This source code is subject to terms and conditions of the Apache License, Version 2.0.
//

using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;

namespace System {

	/// <summary>
	/// Equivalent to the String type in Javascript.
	/// </summary>
	[IgnoreNamespace]
	[Imported(ObeysTypeSystem = true)]
	public sealed class String : IComparable<String>, IEquatable<String> {
		[ScriptName("")]
		public String() {}

		[ScriptName("")]
		public String(String other) {}

		[InlineCode("{$System.Script}.stringFromChar({$System.String}.fromCharCode({ch}), {count})")]
		public String(char ch, int count) {}

		[InlineCode("{$System.String}.fromCharCode.apply(null, {value})")]
		public String(char[] value) {}

		[InlineCode("{$System.String}.fromCharCode.apply(null, {value}.slice({startIndex}, {startIndex} + {length}))")]
		public String(char[] value, int startIndex, int length) {}

		[IndexerName("Chars")]
		public char this[int index] { [InlineCode("{this}.charCodeAt({index})")] get { return '\0'; } }

		[NonScriptable]
		public IEnumerator<char> GetEnumerator() { return null; }

		/// <summary>
		/// An empty zero-length string.
		/// </summary>
		[InlineConstant]
		public const String Empty = "";

		/// <summary>
		/// The number of characters in the string.
		/// </summary>
		[IntrinsicProperty]
		public int Length {
			get {
				return 0;
			}
		}

		/// <summary>
		/// Retrieves the character at the specified position.
		/// </summary>
		/// <param name="index">The specified 0-based position.</param>
		/// <returns>The character within the string.</returns>
		public string CharAt(int index) {
			return null;
		}

		/// <summary>
		/// Retrieves the character code of the character at the specified position.
		/// </summary>
		/// <param name="index">The specified 0-based position.</param>
		/// <returns>The character code of the character within the string.</returns>
		public char CharCodeAt(int index) {
			return '\0';
		}

		[InlineCode("{$System.Script}.compareStrings({s1}, {s2})")]
		public static int Compare(string s1, string s2) {
			return 0;
		}

		[InlineCode("{$System.Script}.compareStrings({s1}, {s2}, {ignoreCase})")]
		public static int Compare(string s1, string s2, bool ignoreCase) {
			return 0;
		}

		[InlineCode("{$System.Script}.compareStrings({this}, {s}, {ignoreCase})")]
		public int CompareTo(string s, bool ignoreCase) {
			return 0;
		}

		[InlineCode("[{s1}, {s2}].join('')")]
		public static string Concat(string s1, string s2) {
			return null;
		}

		[InlineCode("[{s1}, {s2}, {s3}].join('')")]
		public static string Concat(string s1, string s2, string s3) {
			return null;
		}

		[InlineCode("[{s1}, {s2}, {s3}, {s4}].join('')")]
		public static string Concat(string s1, string s2, string s3, string s4) {
			return null;
		}

		/// <summary>
		/// Concatenates a set of individual strings into a single string.
		/// </summary>
		/// <param name="strings">The sequence of strings</param>
		/// <returns>The concatenated string.</returns>
		[InlineCode("{strings}.join('')")]
		public static string Concat(params string[] strings) {
			return null;
		}

		[InlineCode("[{o1}, {o2}].join('')")]
		public static string Concat(object o1, object o2) {
			return null;
		}

		[InlineCode("[{o1}, {o2}, {o3}].join('')")]
		public static string Concat(object o1, object o2, object o3) {
			return null;
		}

		[InlineCode("[{o1}, {o2}, {o3}, {o4}].join('')")]
		public static string Concat(object o1, object o2, object o3, object o4) {
			return null;
		}

		[InlineCode("{o}.join('')")]
		public static string Concat(params object[] o) {
			return null;
		}

		[InlineCode("[{o}].join('')")]
		public static string Concat(object o) {
			return null;
		}

		/// <summary>
		/// Returns the unencoded version of a complete encoded URI.
		/// </summary>
		/// <returns>The unencoded string.</returns>
		[ScriptAlias("decodeURI")]
		public static string DecodeUri(string s) {
			return null;
		}

		/// <summary>
		/// Returns the unencoded version of a single part or component of an encoded URI.
		/// </summary>
		/// <returns>The unencoded string.</returns>
		[ScriptAlias("decodeURIComponent")]
		public static string DecodeUriComponent(string s) {
			return null;
		}

		/// <summary>
		/// Encodes the complete URI.
		/// </summary>
		/// <returns>The encoded string.</returns>
		[ScriptAlias("encodeURI")]
		public static string EncodeUri(string s) {
			return null;
		}

		/// <summary>
		/// Encodes a single part or component of a URI.
		/// </summary>
		/// <returns>The encoded string.</returns>
		[ScriptAlias("encodeURIComponent")]
		public static string EncodeUriComponent(string s) {
			return null;
		}

		/// <summary>
		/// Determines if the string ends with the specified character.
		/// </summary>
		/// <param name="ch">The character to test for.</param>
		/// <returns>true if the string ends with the character; false otherwise.</returns>
		[InlineCode("{$System.Script}.endsWithString({this}, {$System.String}.fromCharCode({ch}))")]
		public bool EndsWith(char ch) {
			return false;
		}

		/// <summary>
		/// Determines if the string ends with the specified substring or suffix.
		/// </summary>
		/// <param name="suffix">The string to test for.</param>
		/// <returns>true if the string ends with the suffix; false otherwise.</returns>
		[InlineCode("{$System.Script}.endsWithString({this}, {suffix})")]
		public bool EndsWith(string suffix) {
			return false;
		}

		/// <summary>
		/// Determines if the strings are equal.
		/// </summary>
		/// <returns>true if the string s1 = s2; false otherwise.</returns>
		[InlineCode("{$System.Script}.compareStrings({s1}, {s2}, {ignoreCase}) === 0)")]
		public static bool Equals(string s1, string s2, bool ignoreCase) {
			return false;
		}

		/// <summary>
		/// Encodes a string by replacing punctuation, spaces etc. with their escaped equivalents.
		/// </summary>
		/// <returns>The escaped string.</returns>
		[ScriptAlias("escape") ]
		public static string Escape(string s) {
			return null;
		}

		[InlineCode("{$System.Script}.formatString({format}, {*values})", NonExpandedFormCode = "{$System.Script}.formatString.apply(null, [{format}].concat({values}))")]
		public static string Format(string format, params object[] values) {
			return null;
		}

		[ExpandParams]
		public static string FromCharCode(params char[] charCode) {
			return null;
		}

		[InlineCode("{$System.Script}.htmlDecode({this})")]
		public string HtmlDecode() {
			return null;
		}

		[InlineCode("{$System.Script}.htmlEncode({this})")]
		public string HtmlEncode() {
			return null;
		}

		[InlineCode("{this}.indexOf({$System.String}.fromCharCode({ch}))")]
		public int IndexOf(char ch) {
			return 0;
		}

		public int IndexOf(string subString) {
			return 0;
		}

		[InlineCode("{this}.indexOf({$System.String}.fromCharCode({ch}), {startIndex})")]
		public int IndexOf(char ch, int startIndex) {
			return 0;
		}

		public int IndexOf(string ch, int startIndex) {
			return 0;
		}

		[InlineCode("{$System.Script}.indexOfString({this}, {$System.String}.fromCharCode({ch}), {startIndex}, {count})")]
		public int IndexOf(char ch, int startIndex, int count)
		{
			return 0;
		}

		[InlineCode("{$System.Script}.indexOfString({this}, {ch}, {startIndex}, {count})")]
		public int IndexOf(string ch, int startIndex, int count)
		{
			return 0;
		}

		[InlineCode("{$System.Script}.indexOfAnyString({this}, {ch})")]
		public int IndexOfAny(params char[] ch) {
			return 0;
		}

		[InlineCode("{$System.Script}.indexOfAnyString({this}, {ch}, {startIndex})")]
		public int IndexOfAny(char[] ch, int startIndex) {
			return 0;
		}

		[InlineCode("{$System.Script}.indexOfAnyString({this}, {ch}, {startIndex}, {count})")]
		public int IndexOfAny(char[] ch, int startIndex, int count) {
			return 0;
		}

		[InlineCode("{$System.Script}.insertString({this}, {index}, {value})")]
		public string Insert(int index, string value) {
			return null;
		}

		[InlineCode("{$System.Script}.isNullOrEmptyString({s})")]
		public static bool IsNullOrEmpty(string s) {
			return false;
		}

		[InlineCode("{this}.lastIndexOf({$System.String}.fromCharCode({ch}))")]
		public int LastIndexOf(char ch) {
			return 0;
		}

		public int LastIndexOf(string subString) {
			return 0;
		}

		public int LastIndexOf(string subString, int startIndex) {
			return 0;
		}

		[InlineCode("{$System.Script}.lastIndexOfString({this}, {$System.String}.fromCharCode({ch}), {startIndex}, {count})")]
		public int LastIndexOf(char ch, int startIndex, int count)
		{
			return 0;
		}

		[InlineCode("{$System.Script}.lastIndexOfString({this}, {subString}, {startIndex}, {count})")]
		public int LastIndexOf(string subString, int startIndex, int count)
		{
			return 0;
		}

		[InlineCode("{this}.lastIndexOf({$System.String}.fromCharCode({ch}), {startIndex})")]
		public int LastIndexOf(char ch, int startIndex) {
			return 0;
		}

		[InlineCode("{$System.Script}.lastIndexOfAnyString({this}, {ch})")]
		public int LastIndexOfAny(params char[] ch) {
			return 0;
		}

		[InlineCode("{$System.Script}.lastIndexOfAnyString({this}, {ch}, {startIndex})")]
		public int LastIndexOfAny(char[] ch, int startIndex) {
			return 0;
		}

		[InlineCode("{$System.Script}.lastIndexOfAnyString({this}, {ch}, {startIndex}, {count})")]
		public int LastIndexOfAny(char[] ch, int startIndex, int count) {
			return 0;
		}

		public int LocaleCompare(string string2) {
			return 0;
		}

		[ExpandParams]
		public static string LocaleFormat(string format, params object[] values) {
			return null;
		}

		public string[] Match(Regex regex) {
			return null;
		}

		[InlineCode("{$System.Script}.padLeftString({this}, {totalWidth})")]
		public string PadLeft(int totalWidth) {
			return null;
		}

		[InlineCode("{$System.Script}.padLeftString({this}, {totalWidth}, {ch})")]
		public string PadLeft(int totalWidth, char ch) {
			return null;
		}

		[InlineCode("{$System.Script}.padRightString({this}, {totalWidth})")]
		public string PadRight(int totalWidth) {
			return null;
		}

		[InlineCode("{$System.Script}.padRightString({this}, {totalWidth}, {ch})")]
		public string PadRight(int totalWidth, char ch) {
			return null;
		}

		[InlineCode("{$System.Script}.removeString({this}, {index})")]
		public string Remove(int index) {
			return null;
		}

		[InlineCode("{$System.Script}.removeString({this}, {index}, {count})")]
		public string Remove(int index, int count) {
			return null;
		}

		[InlineCode("{$System.Script}.replaceAllString({this}, {oldText}, {replaceText})")]
		public string Replace(string oldText, string replaceText) {
			return null;
		}

		[InlineCode("{$System.Script}.replaceAllString({this}, {$System.String}.fromCharCode({oldChar}), {$System.String}.fromCharCode({replaceChar}))")]
		public string Replace(char oldChar, char replaceChar)
		{
			return null;
		}

		[ScriptName("replace")]
		public string ReplaceFirst(string oldText, string replaceText) {
			return null;
		}

		[ScriptName("replace")]
		public string Replace(Regex regex, string replaceText) {
			return null;
		}

		[ScriptName("replace")]
		public string Replace(Regex regex, StringReplaceCallback callback) {
			return null;
		}

		public int Search(Regex regex) {
			return 0;
		}

		public string[] Split(string separator) {
			return null;
		}

		[InlineCode("{this}.split({$System.String}.fromCharCode({separator}))")]
		public string[] Split(char separator) {
			return null;
		}

		public string[] Split(string separator, int limit) {
			return null;
		}

		[InlineCode("{$System.Script}.splitWithCharsAndSplitOptions({this}, {separator})")]
		public string[] Split(char[] separator) {
			return null;
		}

		[InlineCode("{$System.Script}.splitWithCharsAndSplitOptions({this}, {separator}, {options})")]
		public string[] Split(char[] separator, StringSplitOptions options) {
			return null;
		}

		[InlineCode("{this}.split({$System.String}.fromCharCode({separator}), {limit})")]
		public string[] Split(char separator, int limit) {
			return null;
		}

		public string[] Split(Regex regex) {
			return null;
		}

		public string[] Split(Regex regex, int limit) {
			return null;
		}

		[InlineCode("{$System.Script}.startsWithString({this}, {$System.String}.fromCharCode({ch}))")]
		public bool StartsWith(char ch) {
			return false;
		}

		[InlineCode("{$System.Script}.startsWithString({this}, {prefix})")]
		public bool StartsWith(string prefix) {
			return false;
		}

		public string Substr(int startIndex) {
			return null;
		}

		public string Substr(int startIndex, int length) {
			return null;
		}

		public string Substring(int startIndex) {
			return null;
		}

		[ScriptName("substr")]
		public string Substring(int startIndex, int length) {
			return null;
		}

		[ScriptName("substring")]
		public string JsSubstring(int startIndex, int end) {
			return null;
		}

		public string ToLocaleLowerCase() {
			return null;
		}

		public string ToLocaleUpperCase() {
			return null;
		}

		public string ToLowerCase() {
			return null;
		}

		[ScriptName("toLowerCase")]
		public string ToLower() {
			return null;
		}

		public string ToUpperCase() {
			return null;
		}

		[ScriptName("toUpperCase")]
		public string ToUpper() {
			return null;
		}

		public string Trim() {
			return null;
		}

		[InlineCode("{$System.Script}.trimString({this}, {values})")]
		public string Trim(params char[] values)
		{
			return null;
		}

		[InlineCode("{$System.Script}.trimStartString({this}, {values})")]
		public string TrimStart(params char[] values)
		{
			return null;
		}

		[InlineCode("{$System.Script}.trimEndString({this}, {values})")]
		public string TrimEnd(params char[] values)
		{
			return null;
		}

		[InlineCode("{$System.Script}.trimStartString({this})")]
		public string TrimStart() {
			return null;
		}

		[InlineCode("{$System.Script}.trimEndString({this})")]
		public string TrimEnd() {
			return null;
		}

		/// <summary>
		/// Decodes a string by replacing escaped parts with their equivalent textual representation.
		/// </summary>
		/// <returns>The unescaped string.</returns>
		[ScriptAlias("unescape")]
		public static string Unescape(string s) {
			return null;
		}

		[IntrinsicOperator]
		public static bool operator ==(string s1, string s2) {
			return false;
		}

		[IntrinsicOperator]
		public static bool operator !=(string s1, string s2) {
			return false;
		}

		[InlineCode("{$System.Script}.compare({this}, {other})")]
		public int CompareTo(string other) {
			return 0;
		}

		[InlineCode("{$System.Script}.equalsT({this}, {other})")]
		public bool Equals(string other) {
			return false;
		}

		[InlineCode("{$System.Script}.equalsT({a}, {b})")]
		public static bool Equals(string a, string b)
		{
			return false;
		}

		[InlineCode("{args}.join({separator})")]
		public static string Join(string separator, params string[] args)
		{
			return null;
		}

		[InlineCode("{args}.join({separator})")]
		public static string Join(string separator, params Object[] args)
		{
			return null;
		}

		[InlineCode("{$System.Script}.arrayFromEnumerable({args}).join({separator})")]
		public static string Join(string separator, IEnumerable<string> args)
		{
			return null;
		}

		[InlineCode("{$System.Script}.arrayFromEnumerable({args}).join({separator})")]
		public static string Join<T>(string separator, IEnumerable<T> args)
		{
			return null;
		}

		[InlineCode("{args}.slice({startIndex}, {startIndex} + {count}).join({separator})")]
		public static string Join(string separator, string[] args, int startIndex, int count)
		{
			return null;
		}

		[InlineCode("({this}.indexOf({value}) !== -1)")]
		public bool Contains(string value)
		{
			return false;
		}
	}
}
